Sources/Danger/GitLabDSL.swift

```diff
@@ -53,7 +54,12 @@ extension GitLab {
             public let dueDate: Date?
             public let id: Int // swiftlint:disable:this identifier_name
             public let iid: Int
-            public let projectId: Int
+            /// Will has value if it is a regular \
+            /// [project milestone](https://docs.gitlab.com/ee/api/milestones.html). Or it will be nil.
+            public let projectId: Int?
+            /// Will has value if it is a \
+            /// [group milestone](https://docs.gitlab.com/ee/api/group_milestones.html). Or it will be nil.
+            public let groupId: Int?
```

Comment:
What about an enum?

extension GitLab.Milestone {
  enum Identifier: Decodable {
    case groupID(Int)
    case projectID(Int)
  }
}

Reply:
sounds great, let me try it out!

Reply:
You can replace `groupId` and `projectId` with 

```swift
let referenceIdentifier: ReferenceIdentifier
```

With `ReferenceIdentifier` defined like this

```swift
extension GitLab.Milestone {
  enum ReferenceIdentifier: Decodable {
    case groupId(Int)
    case projectId(Int)

     enum CodingKeys: String, CodingKey {
        case groupId
        case projectId
     }

     init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        if let groupId = container.decodeIfPresent(String.self, .groupId) {
           self = .groupId(groupId)
        } else if let projectId = container.decodeIfPresent(String.self, .projectId) {
           self = .projectId(groupId)
        } else {
           throw // Invalid identifier
        }
     }
  }
}
```

Reply:
Thanks for the code!
I was thinking about this at the first, but I was a little bit worried about it will no so noticeable due to the property name is not in the API. How do you think about this?
But I am totally positive to keep the original ones (groupId and project Id) and add this helper-is reference identifier as plus!

Reply:
One of the things I love about Swift is that easily allows to not design impossible states.

In the current `groupId` and `projectId` are both optionals, this creates 4 possible states:
- both `groupId` and `projectId` have a value
- `groupId` has a value
- `projectId` has a value
- none has a value

But those states are not all valid, you only have one between the two values and I believe you always at least one.
I personally prefer an API that differs slightly (but is documented), but makes illegal states impossible, in that way your model gives a good idea of what the user can expect from it.
For a user will be obvious that they are mutually exclusive, given in an enum, without even the need of read the documentation.

(In case you find this interesting and want more details, here there is a really cool talk from objc.io about this https://talk.objc.io/episodes/S01E153-making-impossible-states-impossible?t=104 and one from PointFree https://www.pointfree.co/episodes/ep4-algebraic-data-types :) )

Reply:
Thank you for the information!! It helps a lot. I am using it at view state handling but didn't know it is called impossible state.
It is very reasonable that eliminate the impossible states, I will apply this change and document it to the property.